src/fileio.jl

```diff
@@ -12,30 +13,55 @@ function loadtodict!(d::Dict, g::Union{File,Group}, prefix::String="")
     return d
 end
 
+_set_track_order(kwargs) = if get(kwargs, :track_order, false)
+    prev = IDX_TYPE[]
+    IDX_TYPE[] = HDF5.API.H5_INDEX_CRT_ORDER
+    true, prev
+else
+    false, nothing
+end
+
+_restore_track_order(restore, prev) = (restore && (IDX_TYPE[] = prev); nothing)
+
 # load with just a filename returns a flat dictionary containing all the variables
-function fileio_load(f::FileIO.File{FileIO.format"HDF5"}; kwargs...)
-    h5open(FileIO.filename(f), "r"; kwargs...) do file
-        loadtodict!(Dict{String,Any}(), file)
+function fileio_load(f::FileIO.File{FileIO.format"HDF5"}; _kwargs...)
+    kwargs = Dict{Symbol,Any}(_kwargs)  # mutate `_kwargs`
+    d = pop!(kwargs, :dict, Dict{String,Any}())
+
+    # infer `track_order` from Dict type
+    (track_order = isa(d, OrderedDict)) && get!(kwargs, :track_order, track_order)
+
+    saved = _set_track_order(kwargs)
+    out = h5open(FileIO.filename(f), "r"; kwargs...) do file
+        loadtodict!(d, file)
     end
+    _restore_track_order(saved...)
+    out
 end
 
 # when called with explicitly requested variable names, return each one
 function fileio_load(f::FileIO.File{FileIO.format"HDF5"}, varname::AbstractString; kwargs...)
-    h5open(FileIO.filename(f), "r"; kwargs...) do file
+    saved = _set_track_order(kwargs)
+    out = h5open(FileIO.filename(f), "r"; kwargs...) do file
         read(file, varname)
     end
+    _restore_track_order(saved...)
+    out
 end
 
 function fileio_load(f::FileIO.File{FileIO.format"HDF5"}, varnames::AbstractString...; kwargs...)
-    h5open(FileIO.filename(f), "r"; kwargs...) do file
+    saved = _set_track_order(kwargs)
+    out = h5open(FileIO.filename(f), "r"; kwargs...) do file
         map(var -> read(file, var), varnames)
     end
+    _restore_track_order(saved...)
+    out
 end
 
 # save all the key-value pairs in the dict as top-level variables
 function fileio_save(f::FileIO.File{FileIO.format"HDF5"}, dict::AbstractDict; kwargs...)
     h5open(FileIO.filename(f), "w"; kwargs...) do file
-        for (k,v) in dict
+        for (k, v) in dict
```

Comment:
Shouldn't this also  check
`track_order = isa(d, OrderedDict)`
then pass that into the `h5open` ?

Reply:
I might be confused on the expected outcomes when an `OrderedDict` is saved.

If someone saves an ordered dict, we do not automatically track its order, but shouldn't  we set it automatically?

If the user then reads this back in as a `Dict` it doesn't matter since `Dict` doesn't preserve order. If the user reads in into an `OrderedDict` and they forgot to set `track_order` then we read it back into an `OrderedDict`, but the ordering is not preserved.  Is that correct?

Reply:
Done.

Reply:
> If someone saves an ordered dict, we do not automatically track its order, but shouldn't we set it automatically?
It is now automatic.

Plus, as a safety measure when someone passes an `OrderedDict`, using `get!(...)` ensures that if the user passes `track_order=false` together with and `OD`, then the order won't be tracked in the corresponding `hdf5` file.